merino/providers/weather/provider.py

```diff
@@ -52,14 +74,71 @@ async def initialize(self) -> None:
     def hidden(self) -> bool:  # noqa: D102
         return False
 
+    def cache_key_for_weather_report(self, geolocation: Location) -> Optional[str]:
+        """Compute a Redis key used to look up and store cached weather reports for a location."""
+        cache_inputs = self.backend.cache_inputs_for_weather_report(geolocation)
+        if not cache_inputs:
+            return None
+
+        return f"{self.name}:v1:report:{hashlib.blake2s(cache_inputs).hexdigest()}"
+
+    async def fetch_cached_weather_report(
+        self, geolocation: Location
+    ) -> Optional[WeatherReport]:
+        """Fetch a cached weather report, if available, for a location.
+
+        Raises:
+            CacheMissError: if there's no entry in the cache for this location.
+            CacheEntryError: if the cached weather report can't be deserialized.
+        """
+        cache_key: Optional[str] = self.cache_key_for_weather_report(geolocation)
+        if not cache_key:
+            raise CacheMissError
+
+        cache_value: Optional[bytes] = await self.redis.get(cache_key)
+        if not cache_value:
+            raise CacheMissError
+
+        try:
+            weather_report_dict = json.loads(cache_value)
+            if not weather_report_dict:
+                return None
+            return WeatherReport.parse_obj(weather_report_dict)
+        except ValueError as exc:
+            raise BadCacheEntry("Failed to parse cache entry") from exc
+
+    async def store_cached_weather_report(
+        self, geolocation: Location, weather_report: Optional[WeatherReport]
+    ):
+        """Store a cached weather report, or the absence of one, for a location."""
+        cache_key: Optional[str] = self.cache_key_for_weather_report(geolocation)
+        if not cache_key:
+            return
+
+        cache_value = json.dumps(weather_report.dict() if weather_report else {})
```

Comment:
```suggestion
        cache_value = weather_report.json() if weather_report else "{}"
```
Also, shall we not cache the empty value so that Merino can try fetching from the upstream soon?

Reply:
> shall we not cache the empty value so that Merino can try fetching from the upstream soon?

Great question—I wanted to ask you about that, actually! 😅 `get_weather_report`'s signature suggests that it can succeed but return `None`:

```py
async def get_weather_report(self, geolocation: Location) -> Optional[WeatherReport]
```

And I noticed we have a [test](https://github.com/mozilla-services/merino-py/blob/9a6e0ec40c7d9610a60061b98a17cf42db64355e/tests/unit/providers/weather/test_provider.py#L118-L132) for that behavior. I was wondering if that's intentional—like if AccuWeather doesn't know the location we're requesting, or doesn't have enough data for it?

I opted to explicitly cache the empty value in case that's intentional, because it didn't seem useful to continue sending requests to the upstream if there's no way it can return a report for it. But, if returning `None` means a transient failure, I wonder if we should also change the signature to:

```py
async def get_weather_report(self, geolocation: Location) -> WeatherReport
```

...And have `get_weather_report` raise an error if it can't return a complete report (and not cache that result). WDYT? 😄 (/cc @Trinaa, too)

Reply:
Through another discussion offline, we decided to stick to the current way of caching empty resutls.

Reply:
Sounds good! I'll tweak the comment and docstrings to explicitly mention "negative caching" also.